### PR TITLE
Download DEM for burst bounding box

### PR DESCRIPTION
**Pull request by jtherrmann**
I decided just to copy the file https://github.com/ACCESS-Cloud-Based-InSAR/DockerizedTopsApp/blob/dev/isce2_topsapp/localize_dem.py so that we can discuss what functionality we're attempting to duplicate.

The main function is `download_dem_for_isce2` which writes two files, `full_res.dem.wgs84` and `low_res.dem.wgs84`. It also calls the two helper functions `tag_dem_xml_as_ellipsoidal` and `fix_image_xml` to modify the XML files.

Do we want to preserve all of this functionality, or is there only a subset of the functionality that we're interested in?

Refer to https://github.com/ACCESS-Cloud-Based-InSAR/dem-stitcher/blob/dev/notebooks/Staging_a_DEM_for_ISCE2.ipynb for an example of running the code.

I think the basic steps for finalizing this PR will be:

- [x] Decide what functionality we want to preserve, and remove any unneeded functionality
- [x] Get the code to run successfully on my local machine
- [x] Write tests ~~(will want to refer to https://github.com/ACCESS-Cloud-Based-InSAR/DockerizedTopsApp to see what tests, if any, have already been written for these functions)~~ I see no tests for `download_dem_for_isce2` in the original repo
- [x] Credit the original code appropriately

### CODE REVIEW

**Comment by forrestfwilliams** [comment]
@jtherrmann we will want to keep almost all of the functionality you listed. The helper functions are needed to prep the inputs for ISCE2, but let's remove the low resolution DEM generation code in `download_dem_for_isce2` (lines 85-108). I believe they only create this second DEM as a way to speed up geocoding. We'll add it back in if we feel it is needed.

**Comment by forrestfwilliams** [comment]
Here's an example of the ISCE2 DEM class https://github.com/isce-framework/isce2/blob/8af43d2b9f291e2370e9112a3ead1449d0f2a386/components/isceobj/TopsProc/runVerifyDEM.py#L30

**Comment by jtherrmann** [comment]
> Here's an example of the ISCE2 DEM class https://github.com/isce-framework/isce2/blob/8af43d2b9f291e2370e9112a3ead1449d0f2a386/components/isceobj/TopsProc/runVerifyDEM.py#L30

@forrestfwilliams I looked into using that class for mocking purposes and decided it would take some tinkering to do what we want. I think a more straightforward approach would be to just mock the return values of `stitch_dem` directly, which I've started on [here](https://github.com/ASFHyP3/hyp3-isce2/pull/6/files#diff-db69c1bf5a46ba8ae36921d910763bf1991f7baac46e785cb305ac566bcfb497R10-R29). The next question would be whether we should use a miniature array for the DEM image (which could be written as a literal value in the test file) or a full-sized 3600x3600 array (which would need to be committed as a separate file and then loaded during the test). I also don't know whether it's important for the metadata stored in `DEM_PROFILE` to match the actual size of `DEM_ARRAY`, for example when the DEM is written to disk [here](https://github.com/ASFHyP3/hyp3-isce2/pull/6/files#diff-e459a2a25c131b2d9400df6cea5683f8cdcd0c1294082a418db6ca5988208b5fR107-R108). We can hammer out the details tomorrow morning and then it should be fairly straightforward to finish up the tests and finalize this PR.

**Comment by forrestfwilliams** [comment]
This may also be helpful
https://github.com/isce-framework/isce2/blob/8af43d2b9f291e2370e9112a3ead1449d0f2a386/components/isceobj/Util/ImageUtil/DemImageLib.py

File: src/hyp3_isce2/localize_dem.py
Comment (forrestfwilliams) [inline src/hyp3_isce2/localize_dem.py:1]: I think we should switch the name of this file form `localize_dem.py` -> `dem.py`. The `localize` portion of the name is a holdover from DockerizedTopsApp and doesn't have any meaning here.

**Review by forrestfwilliams — APPROVED** [review]
LGTM! I've added some suggestions that would be good to incorporate, but feel free to merge once you've addressed them.

File: src/hyp3_isce2/localize_dem.py [hunk 1]
```diff
@@ -0,0 +1,110 @@
...
+import rasterio
+from lxml import etree
+from shapely.geometry import box
+
+DEM_RESOLUTION = 0.0002777777777777777775
+
+
+def tag_dem_xml_as_ellipsoidal(dem_path: Path) -> str:
+    xml_path = str(dem_path) + '.xml'
+    assert Path(xml_path).exists()
```
[the remark this answers is not in the document]
Comment (jtherrmann) [reply src/hyp3_isce2/localize_dem.py:31]: Credit goes to the original authors.

File: src/hyp3_isce2/localize_dem.py [hunk 2]
```diff
@@ -0,0 +1,110 @@
...
+    return xml_path
+
+
+def fix_image_xml(xml_path: str) -> None:
+    isce_apps_path = site.getsitepackages()[0] + '/isce/applications'
+    fix_cmd = ' '.join([f'{isce_apps_path}/fixImageXml.py', '-i', xml_path, '--full'])
+    subprocess.check_call(fix_cmd, shell=True)
+
+
+def buffer_extent(extent: list, buffer: float) -> list:
```
[the remark this answers is not in the document]
Comment (jtherrmann) [reply src/hyp3_isce2/localize_dem.py:51]: Took me a while to realize that `wsen` means `west-south-east-north`. Would that be more obvious to GIS people reading the code? I agree the comment is a good idea. I may leave a TODO for this so we can circle back to it.

[the remark this answers is not in the document]
Comment (jtherrmann) [reply src/hyp3_isce2/localize_dem.py]: Actually, the docstr for `download_dem_for_isce2` includes:

```
    extent : list
        [xmin, ymin, xmax, ymax] for epsg:4326 (i.e. (x, y) = (lon, lat))
```

Is that enough to make it clear?

[the remark this answers is not in the document]
Comment (forrestfwilliams) [reply src/hyp3_isce2/localize_dem.py]: Yep, that's good enough!

**Pull request merged**